internal/topic/topicwriterinternal/encoders.go

```diff
@@ -31,22 +59,66 @@ func NewEncoderMap() *EncoderMap {
 				return gzip.NewWriter(writer), nil
 			},
 		},
+		mx: sync.RWMutex{},
+		p:  make(map[rawtopiccommon.Codec]*encoderPool),
 	}
 }
 
-func (e *EncoderMap) AddEncoder(codec rawtopiccommon.Codec, creator PublicCreateEncoderFunc) {
+func (e *MultiEncoder) AddEncoder(codec rawtopiccommon.Codec, creator PublicCreateEncoderFunc) {
 	e.m[codec] = creator
 }
 
-func (e *EncoderMap) CreateLazyEncodeWriter(codec rawtopiccommon.Codec, target io.Writer) (io.WriteCloser, error) {
+func (e *MultiEncoder) Encode(codec rawtopiccommon.Codec, target io.Writer, data []byte) (int, error) {
+	enc, err := e.createEncodeWriter(codec, target)
+	if err != nil {
+		return 0, err
+	}
+
+	n, err := enc.Write(data)
+	if err == nil {
+		err = enc.Close()
+	}
+	if err != nil {
+		return 0, err
+	}
+
+	resetableEnc, ok := enc.(PublicResetableWriter)
+	if ok {
+		e.mx.Lock()
```

Comment:
Here you need two locks every time: for check pool and for put codec. It is difficult to read and to compute

What about create pool for every codec within AddEncoder func?
It allows to remove sync for the maps completely, because them will be changed within single thead on init only.